### PR TITLE
feat: runner implementation & compliance tests

### PR DESCRIPTION
**Pull request by LakiG**
Changes include:

- CLI interface
- TES v1.0 API specs
- YAML test files
- job/test runner
- REST requestor/client
- exception handling
- Unit tests (coverage 95%)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -127,3 +127,6 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+
+# Other
+.idea
```
Comment (uniqueg) [inline .gitignore:132]: It's better to use `.git/exclude` or your global Git ignore for any ignore patterns that are specific to your own setup, like editors etc. The version-controlled `.gitignore` is better suited for stuff that is applicable to _all_ developers. No need to change though, it's really not a big deal. Just FYI

File: cli.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from job_runner import JobRunner
```
Comment (uniqueg) [inline cli.py]: Please order imports in all modules according to [PEP 8](https://peps.python.org/pep-0008/) and [PEP 257](https://peps.python.org/pep-0257/). See this comment for details: https://github.com/elixir-cloud-aai/proTES/pull/74#discussion_r939165612

File: cli.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+def main():
+    pass
+
+
+@main.command(help='Run TES compliance tests against the servers')
+@click.option('--tag', '-t', multiple=True, help='Tag')
+def report(tag):
+
+    if len(tag) == 0:
+        tag = "All",
```
Comment (uniqueg) [inline cli.py]: I would rather have a convention that tags are not case-sensitive, then have the codebase convert all tags to lower-case. It's difficult to remember case conventions, especially mixed case, so it's easier for the user if they can put `ALL`, `all` and `All` to achieve the same thing.

Comment (uniqueg) [reply cli.py]: Also, you can remove the comma from the end of the line here for clarity

File: cli.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    print(f"Input tag is - {tag}")
```
Comment (uniqueg) [inline cli.py]: Better to set up and use the Python `logging` module. See other repos like TRS-Filer, DRS-Filer, proWES etc. to see how to do this in the main entry point (i.e., this module) and then in every other module. Then replace all `print()` function calls with logger calls (e.g., `LOGGER.debug()` or `LOGGER.info()`).

File: compliance_suite/constants/constants.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+from compliance_suite.models.v1_0_specs import TesServiceInfo, TesListTasksResponseMinimal, TesListTasksResponse, \
```
Comment (uniqueg) [inline compliance_suite/constants/constants.py]: Better to use a tuple when importing multiple objects. Also easier for maintenance (e.g., Git diffs) when you put each object on its own line, followed by a comma). E.g.:

```py
from compliance_suite.models.v1.0_specs import (
    TesServiceInfo,
    TesListTasksResponseMinimal,
    ...,
)
```

File: compliance_suite/constants/constants.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    'list_tasks_BASIC': TesListTasksResponse,
+    'list_tasks_FULL': TesListTasksResponse,
+    'get_task_MINIMAL': TesTaskMinimal,
+    'get_task_BASIC': TesTask,
+    'get_task_FULL': TesTask,
+    'create_task': TesCreateTaskResponse,
+    'create_task_request_body': TesTask,
+    'cancel_task': TesCancelTaskResponse
+}
+
```
Comment (uniqueg) [inline compliance_suite/constants/constants.py:45]: So this will always use `v1.0` models, then? Better to find a dynamic solution for this, although I'm fine if this comes in a future commit if you don't have a good idea how to do this quickly (please open an issue for it in that case)

File: compliance_suite/constants/constants.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+REQUEST_HEADERS = {
```
Comment (uniqueg) [inline compliance_suite/constants/constants.py:46]: A useful feature (you can create an issue for that) would be to parameterize this. All requests and responses to/from TES may indeed be JSON at this point, but they may not be in the future (or if you would like to use the runner for other specs)

File: compliance_suite/exceptions/ComplianceException.py [hunk 8]
```diff
@@ -0,0 +1,3 @@
+class ComplianceException(Exception):
+
```
Comment (uniqueg) [inline compliance_suite/exceptions/ComplianceException.py]: Please add a docstring

File: compliance_suite/constants/constants.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+# Utility Constants
+
+COLORS = {
+    'HEADER': '\033[95m',
+    'BLUE': '\033[94m',
+    'CYAN': '\033[96m',
+    'GREEN': '\033[92m',
+    'YELLOW': '\033[93m',
+    'RED': '\033[91m',
+    'END': '\033[0m',
```
Comment (uniqueg) [inline compliance_suite/constants/constants.py]: Might be better called `RESET` (more descriptive)

File: docs/architecture.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      For example, we create a task with "Hello World" command. Now, we will GetTask and validate if the response contains "Hello World" command or not. This can similarly be extended to tackle complex scenarios.
+
+## Report Generator
+
+A final report is generated which showcases all our test results. This report is also send to the central GA4GH Testbed which hosts similar compliance test reports.
+        
+The functions are - 
+1. Generate report - A common GA4GH Testbed Python Library has been developed focusing on generating reports. It will be used to generate the report.
+2. Display HTML web view in local server - As part of CLI arguments, if local server is enabled, a web view will be hosted on the local server which which display our report in HTML.
+3. Push the report to GA4GH Testbed API - To send the report to central GA4GH Testbed, an API request is sent containing relevant info.
```
Comment (uniqueg) [inline docs/architecture.md]: This is a good start. Of course, eventually the runner and the test suite should be independent from one another and live in different repos. Then the docs should be extended to include a detailed spec of the expected YAML file format for the runner to parse/process a given test. People running the test suite would then point to a root directory or repo of test cases containing a set of YAML test specs. Again, something that you can write issues for but no need to do that now.

File: docs/architecture.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+The YAML Test file template is present at `/docs/job_template.yml`
+
+## Components
+
+The compliance suite has 3 components - 
+1. YAML Job Parser
+2. Test Runner
+3. Report Generator
+
+![Architecture Diagram](/docs/images/Architecture.JPG)
```
Comment (uniqueg) [inline docs/architecture.md]: Would be good to have a vector-graphics representation of the diagram rather than JPG, because otherwise it's hard to modify in the future (especially for others)

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
+from pathlib import Path
+from setuptools import (setup, find_packages)
+
+with open('README.md', 'r') as fh:
+    long_description = fh.read()
+
+root_dir = Path(__file__).parent.resolve()
+install_requires = []
```
Comment (uniqueg) [inline setup.py]: I guess you can remove this line. Rather, when you assign to `install_requires` below, you could add a type annotation so that it becomes: `install requires: List = _file.read().splitlines()` (or `list` instead of `List` if the runner uses Python 3.8 or up)

File: setup.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+req = root_dir / 'requirements.txt'
+with open(req, 'r') as _file:
+    install_requires = _file.read().splitlines()
+
+setup(
```
Comment (uniqueg) [inline setup.py:12]: Would be good to add a console entrypoint / executable here. See https://github.com/zavolanlab/htsinfer/blob/dev/setup.py#L34-L38 for example

File: setup.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    name='tes-compliance-suite',
+    version='0.1.0',
+    description='TES Compliance Suite to perform conformance testing to API specs and functionality',
+    long_description=long_description,
+    long_description_content_type='text/markdown',
+    url='https://github.com/elixir-cloud-aai/tes-compliance-suite',
+    author='ELIXIR Cloud & AAI',
+    author_email='alexander.kanitz@alumni.ethz.ch',
+    maintainer='Alexander Kanitz',
+    maintainer_email='alexander.kanitz@alumnni.ethz.ch',
```
Comment (uniqueg) [inline setup.py]: Put yourself as the author and `ELIXIR Cloud & AAI` as maintainer. Use your email address for the `author_email` and [cloud-service@elixir-europe.org](mailto:cloud-service@elixir-europe.org) for `maintainer_email`

File: cli.py (hunk 2 above)
Comment (uniqueg) [inline cli.py]: It's generally standard practice to keep all code in a subdirectory of the repository root. So instead of having `cli.py`, `test_runner.py`, the `compliance_suite` package etc all in the repository root, please move them all to another directory.

File: tests/cancel_task.yml [hunk 15]
```diff
@@ -0,0 +1,37 @@
+name: Cancel Tes Task Job
```
Comment (uniqueg) [inline tests/cancel_task.yml:1]: Better to move all of these to a new repository `tes-compliance-tests` or something. Then rename _this_ repo to something like `ga4gh-test-runner`.
Alos, I'd move the tests of the runner itself to directory `tests/` in the repository root directory (instead of `unittests`) where, in the future, you can also put integration/end-to-end tests etc. This is a more standard setup of a (Python) repo.

File: tests/template/test_template.yml [hunk 16]
```diff
@@ -0,0 +1,36 @@
+name: <Job Name>
```
Comment (uniqueg) [inline tests/template/test_template.yml:1]: There'll probably be comments on the template and the schema later, but I think for now it's fine to just go with what you have. Of course, the template and the schema should remain part of the runner (i.e., _this_) repository and should not be moved to the TES compliance test suite repo (see comment above)

File: compliance_suite/functions/requestor.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
+import json
```
Comment (uniqueg) [inline compliance_suite/functions/requestor.py]: Please add docstrings to modules.

Also, please add docstrings and type hints to all functions/methods in this and all other modules.

File: compliance_suite/functions/requestor.py [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+
+def send_request(server, version, endpoint, id_uri_param, query_params, operation, request_body):
+    if id_uri_param:
+        endpoint = endpoint.replace("{id}", id_uri_param)
+
+    base_url = str(server) + version + endpoint
+    response = None
+    print(f"Sending {operation} request to {base_url}. Query Parameters - {query_params}")
+    if operation == "GET":
+        if query_params:
```
Comment (uniqueg) [inline compliance_suite/functions/requestor.py]: You can probably solve that without a conditional, i.e., find a solution where you can always pass query params (even if none are there). Same goes for the `POST` case.

File: compliance_suite/functions/requestor.py [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+from compliance_suite.exceptions.ComplianceException import ComplianceException
+
+check_cancel = None
+
+
+def send_request(server, version, endpoint, id_uri_param, query_params, operation, request_body):
+    if id_uri_param:
+        endpoint = endpoint.replace("{id}", id_uri_param)
+
+    base_url = str(server) + version + endpoint
```
Comment (uniqueg) [inline compliance_suite/functions/requestor.py]: It's also useful to add type hints for local vars

File: compliance_suite/functions/requestor.py [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+    elif operation == "POST":
+        if request_body:
+            request_body = json.loads(request_body)
+            response = requests.post(base_url, headers=REQUEST_HEADERS, json=request_body)
+        else:
+            response = requests.post(base_url, headers=REQUEST_HEADERS)
+    return response
+
+
+def poll_callback(response):
```
Comment (uniqueg) [inline compliance_suite/functions/requestor.py]: Not sure why you are wrapping up `check_poll()` here...?

File: compliance_suite/functions/requestor.py [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    response = None
+    print(f"Sending {operation} polling request to {base_url}. Query Parameters - {query_params}")
+    if operation == "GET":
+
+        try:
+            response = polling2.poll(lambda: requests.get(base_url, headers=REQUEST_HEADERS, params=query_params),
+                                     step=polling_interval, timeout=polling_timeout,
+                                     check_success=poll_callback)
+        except polling2.TimeoutException:
+            raise ComplianceException(f"Polling timeout for {operation} {base_url}")
```
Comment (uniqueg) [inline compliance_suite/functions/requestor.py]: I find `TimeoutException` to be more explicit/accurate than `ComplianceException`. Good to have custom errors, but they should make sense, semantically. Don't just create a one exception catches all error for an entire application and use it indiscriminately. You can go a long way already with built-in exceptions in most cases.

File: compliance_suite/functions/requestor.py [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+import requests
+from compliance_suite.constants.constants import REQUEST_HEADERS
+import polling2
+from compliance_suite.exceptions.ComplianceException import ComplianceException
+
+check_cancel = None
+
+
+def send_request(server, version, endpoint, id_uri_param, query_params, operation, request_body):
+    if id_uri_param:
```
Comment (uniqueg) [inline compliance_suite/functions/requestor.py]: I find this a bit too TES-specific. Better to abstract that so that it will replace the correct path param with the correct value.

File: compliance_suite/functions/requestor.py (hunk 17 above)
Comment (uniqueg) [inline compliance_suite/functions/requestor.py]: Perhaps more standard to call this module `client.py` rather than `requestor.py`?

File: test_runner.py [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+            # print(request_body)
+
+        if "polling" in job_data.keys():
+
+            check_cancel = False
+            # print(self.auxiliary_space)
+            if "cancel_task" in self.auxiliary_space.keys():
+                check_cancel = True
+
+            response = poll_request(self.server, self.version, job_data["endpoint"],
```
Comment (uniqueg) [inline test_runner.py]: To make your code easier to read, please add the argument names in function calls (at least if there are more than one and it's not a very well known built-in function). It's very difficult to track (and also error-prone) if you pass multiple arguments by position, especially if the functions/methods are not typed

File: compliance_suite/functions/requestor.py [hunk 24]
```diff
@@ -0,0 +1,71 @@
+import json
+import requests
+from compliance_suite.constants.constants import REQUEST_HEADERS
+import polling2
+from compliance_suite.exceptions.ComplianceException import ComplianceException
+
+check_cancel = None
```
Comment (uniqueg) [inline compliance_suite/functions/requestor.py]: I'd rather create a class for the client and pass this in the constructor. Using `global` always makes me a bit uneasy.

File: job_runner.py [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+from compliance_suite.functions.colored_console import print_blue, print_red, print_green, print_yellow
+
+
+class JobRunner():
+
+    def __init__(self, tags):
+        self.path = os.getcwd()
+        self.tags = tags
+        self.test_count = 0
+        self.result = {             # To store the count of passed/failed/skipped tests
```
Comment (uniqueg) [inline job_runner.py]: I think it'd be better to provide a method that calculates the numeric results on the fly by processing `.test_status`. No need to store and update this information separately

File: job_runner.py [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+from compliance_suite.exceptions.ComplianceException import ComplianceException
+from compliance_suite.functions.colored_console import print_blue, print_red, print_green, print_yellow
+
+
+class JobRunner():
+
+    def __init__(self, tags):
+        self.path = os.getcwd()
+        self.tags = tags
+        self.test_count = 0
```
Comment (uniqueg) [inline job_runner.py]: Could also be obtained from `.test_status` (see next comment)

File: job_runner.py [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+    def validate_yaml(self, file, yaml_file):
+        try:
+            yaml_data = yaml.safe_load(file)
+            return yaml_data
+        except yaml.YAMLError as err:
+            raise ComplianceException(f"Invalid YAML File - {yaml_file}. Error details - {err}")
+
+    def run_jobs(self):
+
+        yaml_path = os.path.join(self.path, "tests")
```
Comment (uniqueg) [inline job_runner.py]: Better to parametrize this, i.e., pass the entire folder with tests as an argument to the constructor. Should take this from CLI.

File: job_runner.py [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+        print_yellow("#{:^88}#".format(""))
+        print_yellow("#{:^88}#".format(f"Total Tests - {self.test_count}"))
+        print_yellow("#{:^88}#".format(f'Passed - {self.result["passed"]} ({passed_tests})'))
+        print_yellow("#{:^88}#".format(f'Failed - {self.result["failed"]} ({failed_tests})'))
+        print_yellow("#{:^88}#".format(f'Skipped - {self.result["skipped"]} ({skipped_tests})'))
+        print_yellow("#{:^88}#".format(""))
+        print_yellow("{:#^90}".format(""))
+
+    def validate_job(self, yaml_data, yaml_file):
+        schema_path = os.path.join(self.path, "tests", "template", "test_template_schema.json")
```
Comment (uniqueg) [inline job_runner.py]: Might as well pass the entire path to the template file as an argument to the constructor.

File: job_runner.py [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+import yaml
+from test_runner import TestRunner
+from compliance_suite.exceptions.ComplianceException import ComplianceException
+from compliance_suite.functions.colored_console import print_blue, print_red, print_green, print_yellow
+
+
+class JobRunner():
+
+    def __init__(self, tags):
+        self.path = os.getcwd()
```
Comment (uniqueg) [inline job_runner.py]: Replace with `schema_path` and add another arg `tests_dir` - see comment belows

File: job_runner.py [hunk 30]
```diff
@@ -0,0 +1,97 @@
+import os
```
Comment (uniqueg) [inline job_runner.py]: Better to use `pathlib` than `os`

File: job_runner.py [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+        with open(schema_path, "r") as f:
+            json_schema = yaml.safe_load(f)
+
+        try:
+            validate(yaml_data, json_schema)
+            print(f'Test YAML file valid for {yaml_file}')
+        except ValidationError as err:
+            raise ComplianceException(f"YAML schema validation error - {yaml_file}. Error details - {err.message}")
+
+    def validate_yaml(self, file, yaml_file):
```
Comment (uniqueg) [inline job_runner.py]: This function is essentially a single statement (`.safe_load`) that you only call once. Might as well integrate in `.run_jobs()`, especially since you catch the error again there.

File: job_runner.py [hunk 32]
```diff
@@ -0,0 +1,97 @@
...
+
+        yaml_path = os.path.join(self.path, "tests")
+        for yaml_file in os.listdir(yaml_path):
+            if yaml_file.endswith(".yml"):
+                self.test_count += 1
+                with open(os.path.join(yaml_path, yaml_file), "r") as f:
+                    print("\n{:#^100}".format(f"     Initiating Test-{self.test_count} for {yaml_file}     "))
+                    try:
+                        yaml_data = self.validate_yaml(f, yaml_file)
+                        self.validate_job(yaml_data, yaml_file)
```
Comment (uniqueg) [inline job_runner.py]: No need to do all that with the file open. You can put the `catch` right after the YAML load call, then go back two levels of indentation once you have `yaml_data`.

File: job_runner.py [hunk 33]
```diff
@@ -0,0 +1,97 @@
...
+                                test_runner = TestRunner(yaml_data["server"], yaml_data["version"][0])
+                                job_count = 0
+                                for job in yaml_data["jobs"]:
+                                    job_count += 1
+                                    print(f'Running tests for sub-job-{job_count} -> {job["name"]}')
+                                    test_runner.run_tests(job)
+                                self.result["passed"] += 1
+                                self.test_status["passed"].append(str(self.test_count))
+                                print_green(f'Compliance Test-{self.test_count} for {yaml_file} successful.')
+                        if not tag_matched:
```
Comment (uniqueg) [inline job_runner.py]: Not sure what you need `tag_matched` for. You could basically just use an `else` block for `if tag in yaml_data["tags"]`

File: job_runner.py [hunk 34]
```diff
@@ -0,0 +1,97 @@
...
+                            self.result["skipped"] += 1
+                            self.test_status["skipped"].append(str(self.test_count))
+                            print_blue(f"No Tag matched. Skipping Test-{self.test_count} for {yaml_file}")
+
+                    except ComplianceException as err:
```
Comment (uniqueg) [inline job_runner.py]: Not so good to have a lot of statements behind `try` - and have a blanket `catch` for all of them. Users will have a hard time telling what went wrong. Seems to me that only a few of the statements in the `try` block would be able to raise a `ComplianceException`.

**Review by uniqueg — CHANGES_REQUESTED** [review]
Thanks a lot, looks great. Got quite a few suggestions, mostly high level. There are probably quite a few other things that could be improved, but we can do that later.

File: job_runner.py (hunk 33 above)
Comment (LakiG) [reply job_runner.py]: We are validating against multiple user-provided tags. Using if-else would give result for only the first tag. So, I am using the `tag_matched` to check if no tags were matched.

File: job_runner.py (hunk 34 above)
Comment (LakiG) [reply job_runner.py]: I am catching all the exceptions under one big exception because we have multiple test files, so we run through all files via loop. If a file gives exception, I should still continue on remaining files. So, the exception will just print some error message and update test status instead of exiting the code. The ways to do this are - 
1. Use `continue` while catching, but we cant use this for calling methods, as they are outside loop (eg. .validate_job() or the Test Runner methods). 
2. I use a big exception to catch all of them and print the error in this big exception (i.e. `ComplianceException`)

Please suggest if I can take use any other way

File: .gitignore (hunk 1 above)
Comment (LakiG) [reply .gitignore:132]: Sure. I will take care now on.

File: compliance_suite/constants/constants.py (hunk 6 above)
Comment (LakiG) [reply compliance_suite/constants/constants.py:45]: I started with the `v1.0` model. I will add `v1.1` later when things are smooth. When we have multiple versions, we will have to mapping for each model, which will be decided via YAML file version parameter.

File: docs/architecture.md (hunk 10 above)
Comment (LakiG) [reply docs/architecture.md]: Definitely. I will start working on it after this PR, as we plan to expand the compliance suite functionality to other GA4GH services as well. I will raise an issue for this to track it.

File: setup.py (hunk 14 above)
Comment (LakiG) [reply setup.py]: Sure. Thank you

File: tests/cancel_task.yml (hunk 15 above)
Comment (LakiG) [reply tests/cancel_task.yml:1]: When we will split the repo into runner and tests, I will do these changes

File: job_runner.py (hunk 27 above)
Comment (LakiG) [reply job_runner.py]: Passing as argument seems a better approach as it had become messy to handle the file paths. However, after we split the repo, we will extract the files from "tests" repo. Can we leave this for later as it will be updated then.

**Comment by LakiG** [comment]
I have updated the code for most of the comments. Thank you for suggesting them, some of them really improved the code structure.
Just 2 major things remaining and then I will push the code - 
1. Regarding File Paths - Since we are splitting the repo, so it will be updated later.
2. Regarding `ComplianceException` - I am catching all exceptions under it so that I can process all the Test files without exiting the code.

File: job_runner.py (hunk 26 above)
Comment (LakiG) [reply job_runner.py]: We will require this variable to print which test no. is being run in the logs and update the test status.

Comment (uniqueg) [reply job_runner.py]: Okay, fair enough

File: job_runner.py (hunk 27 above)
Comment (uniqueg) [reply job_runner.py]: Sure. Just add these things to the issue description (or a comment) so that they won't be forgotten

File: job_runner.py (hunk 33 above)
Comment (uniqueg) [reply job_runner.py]: Right. My bad. I still think this could be made simpler though.

How about putting the tag matcher in a separate method `.tag_matcher()` that takes a set of tags (from `yaml_data["tags"]`) as input and returns a `bool`? It's a very well defined problem and would benefit from encapsulation. If we change the tag matching logic later for some reason, the behavior of the `.run_jobs()` method (and its unit tests) wouldn't be affected.

File: job_runner.py [hunk 35]
```diff
@@ -0,0 +1,97 @@
...
+                        self.result["failed"] += 1
+                        self.test_status["failed"].append(str(self.test_count))
+                        print_red(f'Compliance Test-{self.test_count} for {yaml_file} failed.')
```
Comment (uniqueg) [inline job_runner.py]: By the way, here's a possible solution to use `logging` with colors: https://stackoverflow.com/a/56944256

Comment (uniqueg) [reply job_runner.py]: Or, perhaps better, use one of the following libraries (examples in the StackOverflow link above):
* [rich](https://pypi.org/project/rich/)
* [colorlog](https://pypi.org/project/colorlog/)
* [coloredlogs](https://pypi.org/project/coloredlogs/)

File: job_runner.py (hunk 34 above)
Comment (uniqueg) [reply job_runner.py]: Okay, I get it now. But still I have a couple of comments:

1. `ComplianceException` can be misleading. If a server fails a more functional test, e.g., a TES instance is unable to deal with `s3`, it's not a compliance issue, at least not necessarily. How about `TestFailureException`?
2. I don't think that lumping together job validation issues (i.e., problems with the YAML files) with compliance/functional issues of the tested implementations makes a whole lot of sense. If the test files are buggy, it's not the fault of the service being tested, and it's not their responsibility to fix them. So I think those two types of problems should be separated. You could define `JobValidationException` for job validation issues. You may also want to define `TestRunnerException` for issues of the the test runner exception itself (although you can add that at a later point, once the logic is more stable). 
3. Due to what I wrote in the last point, I think that solution (1) you are suggesting, i.e., wrapping only the `test_runner.run_tests(job)` statement in a `try` block and amending the `except` block with `continue, indeed makes more sense here for test failure issues. Then catch the validation errors in a different block and handle accordingly.

File: test_runner.py [hunk 36]
```diff
@@ -0,0 +1,104 @@
...
+            endpoint_model = self.job_data["name"] + "_" + view_query[0]
+        else:
+            endpoint_model = self.job_data["name"]
+        self.validate_logic(endpoint_model, response_json, "Response")
+        self.auxiliary_space[self.job_data["name"]] = response_json
+
+    def run_tests(self, job_data):
+
+        # print(job_data)
+        self.job_data = job_data
```
Comment (uniqueg) [inline test_runner.py]: Setting instance attributes inside a non-setter method is potentially dangerous. You are doing this both here, for `self.job_data`, and above, for `self.auxiliary_space`. Apart from the usual pitfalls of side effects, this makes it quite hard to follow the code (especially if attributes and methods are not documented). I would feel much more comfortable if you used the setter pattern if you MUST modify instance attributes. But even better if you can refactor in a way that you can pass these as params to dependent functions.

File: job_runner.py (hunk 34 above)
Comment (LakiG) [reply job_runner.py]: I have made 3 Exceptions in the suite. `JobValidationException`, `TestFailureException`, `TestRunnerException`. The last one is not used currently, we will add it later.

Regarding that `continue` approach, except the `.validate_yaml()`, all the errors have to be caught under `JobValidationException` & `TestFailureException`. We will be writing the whole `continue` code for just 1 function, which can also be caught under `JobValidationException`.

Should we skip the `continue` method then? So basically, the changes would be, instead of using `ComplianceException` where we do not get a clear message about the exception, we split into 3 well-defined exceptions (`JobValidationException`, `TestFailureException` & `TestRunnerException`), but they will be caught in the same way as before.

File: test_runner.py (hunk 36 above)
Comment (LakiG) [reply test_runner.py]: Passing the instance attributes as params can be difficult as there are too many values to handle.
I will use the setter method. Though I will not use the instance attribute. I will just set the instance attribute into class attribute. Then onwards, I will use class attribute.

File: job_runner.py (hunk 34 above)
Comment (uniqueg) [reply job_runner.py]: Let's discuss this on the call now.

**Review by uniqueg — APPROVED** [review]
